File: 15 3Sum/15 LL.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

/*
https://leetcode.com/problems/3sum/

Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0?
Find all unique triplets in the array which gives the sum of zero.

Note:
Elements in a triplet (a,b,c) must be in non-descending order. (ie, a ≤ b ≤ c)
The solution set must not contain duplicate triplets.
For example, given array S = {-1 0 1 2 -1 -4},

A solution set is:
(-1, 0, 1)
(-1, -1, 2)
*/

class Solution
{
public:
	vector<vector<int> > threeSum(vector<int>& nums)
	{
		vector<vector<int> > result;
		if (nums.size() < 3) return result;
		sort(nums.begin(), nums.end());
		vector<int>::iterator first = nums.begin();
		int val_first = ~first[0];	// memorize
		for (; first != nums.end(); ++first)
		{
			// remove duplicate
			if (val_first == first[0]) continue;
			val_first = first[0];
			vector<int>::iterator second = first + 1;
			vector<int>::iterator third = nums.end() - 1;
			// typical two slide pointer
			while (second < third)
			{
				if ((val_first + second[0] + third[0]) < 0)
				{
					++second;
				}
				else if ((val_first + second[0] + third[0]) == 0)
				{
					vector<int> triplet(3);
					triplet[0] = val_first;
					triplet[1] = second[0];
					triplet[2] = third[0];
					result.push_back(triplet);
					// remove duplicate
					while ((second < third) && (second[0] == triplet[1]))
					{
						++second;
					}
				}
				else
				{
					--third;
				}
			}
		}
		return result;
	}
};

/*
idea: loop first value, slide two additional values.
complexity: Time O(N^2) Space O(1)
*/

int main(void)
{
	vector<int> nums;
	nums.push_back(-2);
	nums.push_back(-1);
	nums.push_back(-1);
	nums.push_back(1);
	nums.push_back(1);
	nums.push_back(2);
	Solution engine;
	cout << engine.threeSum(nums).size() << '\n';
	return 0;
}

File: 15 3Sum/15 LL.py
class Solution(object):
	def threeSum(self, nums):
		result = list()
		if(len(nums) < 3):
			return result
		nums.sort()
		val_a = ~nums[0]
		for a in range(0, len(nums)):
			if val_a == nums[a]:
				continue
			val_a = nums[a]
			b = a + 1
			c = len(nums) - 1
			while b < c:
				if val_a + nums[b] + nums[c] < 0:
					b += 1
				elif val_a + nums[b] + nums[c] == 0:
					result.append([val_a, nums[b], nums[c]])
					val_b = nums[b]
					while b < c and val_b == nums[b]:
						b += 1
				else:
					c -= 1
		return result

engine = Solution()
print engine.threeSum([2, 0, -2])